### ISSUE TITLE
Create Players from Audio Element from HTML

### ISSUE THREAD

**Issue by X-Raym** [katspaugh/wavesurfer.js#524]
Hi,

Question:
Tell me if I am wrong, but at this point, all datas are generated from a the jquery file.
Would it be possible to tweak it so that it is able to auto-detect audio element on HTML (or maybe a div or something else with data attributes) and put them this player ?
The ultimate goal would be to be able to auto create player from the html without touching the JS, and if possible, creating some kind of wordpress plugin with wavesurfer.
The fact all infos came from the jquery make it a bit difficult to integrate with wordpress from post content.

Thanks for the infos !

**Comment by katspaugh** [comment]
Good idea, we should add automatic HTML initialization.

**Comment by X-Raym** [comment]
Oh, I'm very happy that you are interested by this :)

This would be wonderful and so much easier ton integrate in post content.
Idealy, the markup should be abble to determine some customization (such as the plugins to used for a certain file player for exemple - multi waveforms, etc...), and it would fall back to some default settings if nothing provided.

this is a lot of work in perspective but IMHO it really worth it.

thanks for you interest in this, can't wait to hear more about it !

**Comment by tweed** [comment]
VERY glad to read 'HTML initialization" -- I use video.js for video and am committed to wavesurfer.js ... and, someday figuring out (or benefiting from new HTML initialization:-) HOW I may have an audio player in every article (html/css/js only pages) without putting a script for each player etc. etc.

So far I've been effectively stopped as the few examples I've found in 'issues' posts seem to only hint at some elegant method. 

I'm relentlessly optimistic - and yet, IS HTML initialization the key?

For video.js, of course, I simply put HTML calling the .mp4/.ogg/.webm and done.

Is it possible, anytime now *-) something like this will appear, allowing, what must be 1000's of us hopefuls, to use HTML, call the audio file, and have a different waveform in each article.

Again, me, i use just plain html/css/js -- no CMS or similar.

In absolutely any event great thanks for all the brilliances.

**Comment by X-Raym** [comment]
closed ? is that integrated or don't you want to implement this anymore ?

**Comment by tweed** [comment]
WOW - I went to: 
@katspaugh 	add html init; resolves #524
as noted above.
Then found a totally new file i believe:
 32  src/html-init.js

And LOTS of changes in other places.
Terribly sorry to have to ask.
Can we find out how to use the new feature?  
(may be an update at some point in readme i figure)

AMAZING body or work. I'm certain we all appreciate the effort and knowledge and creativity,  and ...ummm, patience it takes!

**Comment by katspaugh** [comment]
Here's a codepen with a demo, guys: http://codepen.io/katspaugh/pen/WQXzpB?editors=10
(Didn't post it at once because had to to go right after I commited it.)

**Comment by tweed** [comment]
Let me say: WHOA WHOA WOOHOO!!!!   Changes Everything from my view. Better stop writing as I could ramble on!  HUGE Gratitude.

**Comment by X-Raym** [comment]
@katspaugh 
That's very promising, seems simple to integrate !
few questions : are you sure wavesufer markup won't be a flag as invalid HTML 5 properties ? It could be bad for SEO.
Could some styling be applied by CSS ?

Anyway, very nice to see this update, it is definitely a huge thing :)

@tweed 
You are welcome ^^

**Comment by katspaugh** [comment]
Thanks for the motivation @tweed! :thumbsup:

@X-Raym `<wavesurfer>` is not a valid tag, yes. You think, it can be important for SEO? We can change it to match any element with a `data-wavesurfer` attribute.

You can apply CSS, yes.

**Comment by oaeide** [comment]
Thanks for you amazing work with this! 

Will this change perhaps make it possible to use wavesurfer with meteorjs? Never managed to get it to work before..

**Comment by X-Raym** [comment]
@katspaugh
ok so I did a bit more research on custom tags in HTML5 and it seems possible to use wavesurfer tag without problem, as it is part of W3C recommandations.
http://w3c.github.io/webcomponents/spec/custom/
that said, it seems that it is just 'draft'.
another great article on that:
http://www.html5rocks.com/en/tutorials/webcomponents/customelements/?redirect_from_locale=fr

if after taking a look at this article it seems possible to make <wavesurfer> valid, so yes, go for it, it is a nice idea :)
else, maybe just using a standard <object> is enough.

thanks for your consideration !


ps: could the html markup be able to trigger wavesurfer module such as the split waveforms for multichannel audio ?

**Comment by katspaugh** [comment]
> ould the html markup be able to trigger wavesurfer module such as the split waveforms for multichannel audio ?

Multichannel audio is not a module, it's just a parameter. As I said, all parameters are supported as `data`-attributes.

**Comment by X-Raym** [comment]
@katspaugh
ok, that is perfect :D good job on this feature, it will be so easier to implement :P

**Comment by X-Raym** [comment]
@katspaugh 
Hi,
Does the HTML validation still need the wavesufer tag or is their now a more W3C compliant solution ? :)

**Comment by X-Raym** [comment]
People interested in this can take a look at this WordPress integration project:
[WaveSurfer-WP — A WordPress Integration for WaveSurfer](https://github.com/katspaugh/wavesurfer.js/issues/590)

**Comment by katspaugh** [comment]
> Does the HTML validation still need the wavesufer tag or is their now a more W3C compliant solution ? :)

@X-Raym yep. I still think it's nicer than having a div with some attribute. Besides, wavesurfer.js creates `wave` and `region` tags inside anyway, and I would prefer to keep things as they are.

**Comment by tweed** [comment]
[QUOTE]
Here's a codepen with a demo, guys: http://codepen.io/katspaugh/pen/WQXzpB?editors=100
[END QUOTE]

Sorry to be stumped at such a beginner's level.
The codepen quoted above works. There's no JS or CSS in the codepen. How is it running?

I am only needing to create the most basic WAV or mp3 player multiple times on single page website.
Each blog entry on the single page is within a DIV.
HUGELY great is the new enhancement: 'Create Players from Audio Element from HTML"


I put THIS in HEAD of single webpage

        <script src="wavesurfer.min.js"></script>
		<script src="html-init.js"></script>
		
        <link rel="stylesheet" href="style.css" />
        <link rel="stylesheet" href="ribbon.css" />

THEN I thought I could add THIS to each DIV and get multiple players
<wavesurfer
            data-wave-color="violet"
            data-progress-color="purple"
            data-url="http://ia801502.us.archive.org/25/items/spc148_1510_librivox/spc148_toamerica_sabt_128kb.mp3">
</wavesurfer>


May I ask:
what's the minimum i need to have on the single page of HTML/CSS to have a different wav/mp3 in each DIV (each div is a separate blog entry)

DO I need to add the "controls" HTML in EACH DIV?
AND/OR do i need to add a script to the bottom of the single webpage?

BASICALLY I'm trying to use the HTML initialize and play enhancement with as little repeated code as possible. A different wavesurfer instance in each DIV - all on one single HTML/CSS page.

Sorry to be at the point I need to ask.
Looked everywhere i could for examples.
Tried the usual dozens of attempts.

**Comment by X-Raym** [comment]
@tweed 
When building [WaveSurfer-WP](https://wordpress.org/plugins/wavesurfer-wp/), I realized that we could create WaveSurfer elements from HTML, but using the regular WebAudio just like regular wavesurfer-js initialization with a bit of jQuery.

To make this work with several players, I recreate all buttons elements and the JS initiliazer.
(oh yes and codepen don't need to include scripts as you did, CSS and JS are automatically loaded from the different panels, and external files like wavesurfer-js are put in the settings as external resources).

So, if you need a HTML markup + initializer demo,
Maybe take a look at this [pen](http://codepen.io/X-Raym/pen/VemqzZ)
This is what I have done for WaveSurfer-WP so far.
Contrary to standard HTML init, it doesn't recognize all data attributes, but it has smoother cursor refresh rate.

Feel free to mod it to your needs !

**Comment by tweed** [comment]
WOW  X-Raym !! thank YOU

I'm usually able to stumble upon the answer through searches and attempts - not so with wavesurfer - so far.

Before posting and asking here I had already downloaded and looked at your fab WP plugin and also had looked at your pen.

Having failed with 'html-init.js' before posting the above, upon seeing your codepen's jquery code, and your plugin's PHP code I only did a quick look.  My single page HTML/CSS/JS blog does not use jquery, though I will be doing a WP effort as well and your plugin will be FANTASTIC. I'll be in good company there *-) nice!

Did a few effort to finding out what the external resources are for codepens that don't show them. no luck.
Firstly: I thought that, if I could find out what JS and CSS was behind: http://codepen.io/katspaugh/pen/WQXzpB?editors=100
i'd know if I'm on the right track in hoping to have as minimal/vanilla as possible multi-waversurfer instances on one simple HTML/CSS/JS page.


'html-init.js' would likely be best to start with as i think I'll need some of the attributes.

Secondly, my wishful thinking is:

1) put wavesurfer.js and html-init.js in HEAD of single HTML/CSS/JS page.

2) perhaps some CSS needs to be in the HEAD as well - ribbon.css? bottom part of wavesurfer's style.css?

3) each blog post has 3 DIVs - one for text and video, and audio. Perhaps needed would be a single bit of JS script on the page at the bottom with the other JS scripts. 

4) leave off the button HTML code (at least NOT have to have it in EVERY DIV) - and leave off any bootstrap as I only want a basic non-styled PLAY/PAUSE single button and the waveform.

super minimal -  i thought possibly the HTML initialization enhancement would allow very little code needed in each audio DIV - as, there are many many blog posts on the single HTML page.

wavesurfer
            data-wave-color="violet"
            data-progress-color="purple"
            data-url="media/blogpost1.mp3">
/wavesurfer>

wavesurfer
            data-wave-color="violet"
            data-progress-color="purple"
            data-url="media/blogpost2.mp3">
/wavesurfer


NOTE:  the < and the > get deleted from the above, around 'wavesurfer'

**Comment by X-Raym** [comment]
External resources on codepen are under the (Settings)[https://monosnap.com/file/rUsyln3cAKiS3fZXYrX3FEWU2ecdb3.png] panel.
You only need the wavesurfer.min.js file. These are the basic resources.
Then you can have your own CSS and your own initialization javascript. It is easier to build a initialization file with jQuery because of all the event detector and DOM manipulation.
I don't mind using jQuery as it is bundle on all WordPress installation.

Hope these infos help !

(consider using code markup when posting code so that it become more readable :hamster: )

**Comment by tweed** [comment]
THANK YOU
-- found the external resources
-- have looked, and haven't found how to post code yet -- sorry!
-- jquery's great - just keen to not use it with my simple vanilla js page so i guessi need to find out how to initialize without it.

I've failed entirely to get anything to even show - perhaps it's that I'm running locally only or some other typical silliness - I've learned lots that hard long way *-)

May I ask, can i do this? [below]

have html-init.js and wavesurfer.js in HEAD of web page
have wavesurfer's style.css in HEAD of web page

put wavesurfer tag on webpage ANYWHERE
and have it show up and run?
JUST this tag in HTML page: (as html-init.js saysit needs wavesurfer tag)
<wavesurfer
            data-wave-color="violet"
            data-progress-color="purple"
            data-url="http://ia801502.us.archive.org/25/items/spc148_1510_librivox/spc148_toamerica_sabt_128kb.mp3">
</wavesurfer>

IF so then it would be something like me running locally or other silliness why i get nothing showing.
However, I'm guessing I'm missing something fundamental.
Sorry to be at the point where I really need to sort this out.

**Comment by tweed** [comment]
ADDITIONAL:
I don't get any waveform when i see this codepen:

 http://codepen.io/katspaugh/pen/WQXzpB?editors=100

I am very happy to have had help here.
guessing that as this is a closed "issue' i could/should really consider well, and then post a clear/short (with code done properly) and with example pen once I've done a serious effort at refining my thoughts/observations.

That way it could be helpful to the community.
I believe that people are struggling with the html-init - Multiple Players from Audio Element from HTML  method/option.

Thanks!

**Comment by tweed** [comment]
I'm studying: and I'm now thinking that partly the reason nothing shows is that I need to specify "container' From the WaveSurfer Options

container:	 
CSS-selector or HTML-element where the waveform should be drawn. This is the only required parameter.


If I specify 'container' it would 'draw' based on just an HTML-element

I will sort this out, though if an example is around based on my post 2 above this that would be great of course -- Yet, in absolutely any event - I'm most happy for the time and effort - THANKs!

**Comment by X-Raym** [comment]
In order to help you understand a bit more what is going on, I tell you some infos, hoping it will help you.

In fact, there is two concepts :

1. HTML init from a wavesurfer tag
1. HTML init from a regular tag

The first init, that was discusting about in this thread, is a way to automatically replace wavesurfer tags with a colored waveform, based on all data-thing attribute of this tag.
It also put an audio tag, from which the sound came from. Here is a (demo)[http://codepen.io/X-Raym/pen/jbRawz/]. Look the wavesurfer tag, and all check for the audio tag in the js.

The @katspaugh exemple however didn't have buttons controls for that: indeed, these are made a custom initiliazer-js, different from projects to projects to user needs.

In WaveSurfer-WP v1.x, I made button to control the play of the audio tag.

But there was an issue regarding to refresh rate. Indeed, "new position" events trigger of the audio tags are in fact a bit slow.

So @katspaugh leads me to another concept : using the regular WebAudio element used by wavesurfer-js in default mod.

This needs to change the wavesurfer tag by a div tag (in order to avoid the HTML init using audio element), and then same thing : rebuilding audio controller according to your needs.
I keep the data-attributes concept to push parameter inside my initialization-js, as it is HTML 5 and jQUery friendly.

In both cases I used jQuery for ease of programming (ability to trigger/check events, to look for elements in DOM etc).

The results looks exactly the same, but this WebAudio initlization has a lot of advantage (including nice cursor refreshing rate. Results is just like the first codepen I send you. No wavesurfer tag, and buttons control a WebAudio elements.

I think that is how much I can help you. You have all infos in these two codepens according to your need. I will advice use the WebAudio + jQuery version :+1: 

Feel free to mod everything to your needs.

Good luck for the next !

**Comment by tweed** [comment]
FANTASTIC!
biGGest thanks!

**Comment by sanjeev441** [comment]
Hello @katspaugh,

I used wavesurfer.js package its working great.

I want to know, can we differentiate voice from audio file using wavesurfer.js.
in audio conversion between two person can we separate Two Voice's showing different color?
Please give me replay  as soon as possible.

**Comment by X-Raym** [comment]
@sanjeev441
Please open a dedicated issue, it has nothing to do with this one.

**Comment by sanjeev441** [comment]
do you know any voice differentiate or separate packages.
